src/SA/listmods/entry.c

```diff
@@ -0,0 +1,130 @@
+  LPCSTR  lpSubBlock,
+  LPVOID  *lplpBuffer,
+  PUINT   puLen
+);
+
+// Inspired from: https://www.codeguru.com/cpp/w-p/win32/versioning/article.php/c4539/Versioning-in-Windows.htm
+int PrintSingleModule(char* szFile){
+    DWORD dwLen, dwUseless;
+    LPTSTR lpVI;
+    char* companyName;
+
+    GetFileVersionInfoSizeA_t* GetFileVersionInfoSizeA = (GetFileVersionInfoSizeA_t*) GetProcAddress(LoadLibraryA("Api-ms-win-core-version-l1-1-0.dll"), "GetFileVersionInfoSizeA");
+    dwLen = GetFileVersionInfoSizeA((LPTSTR)szFile, &dwUseless);
+    if (dwLen==0){
+        internal_printf("%-50sERROR: Could not GetFileVersionInfoSizeA() on the DLL.\n", szFile);
+        return 1;
+    }
+
+    lpVI = (LPTSTR) KERNEL32$GlobalAlloc(GPTR, dwLen);
+    if (lpVI) {
+        DWORD dwBufSize;
+        VS_FIXEDFILEINFO* lpFFI;
+        BOOL bRet = FALSE;
+        WORD* langInfo;
+        UINT cbLang;
+        char szVerDescription[128];
+        char szVerCompanyName[128];
+        LPVOID lpDescription;
+        LPVOID lpCompanyName;
+        UINT cbBufSize;
+
+        GetFileVersionInfoA_t* GetFileVersionInfoA = (GetFileVersionInfoA_t*) GetProcAddress(LoadLibraryA("Api-ms-win-core-version-l1-1-0.dll"), "GetFileVersionInfoA");
+        GetFileVersionInfoA((LPTSTR)szFile, 0, dwLen, lpVI);
+
+        //First, to get string information, we need to get
+        //language information.
+        VerQueryValueA_t* VerQueryValueA = (VerQueryValueA_t*) GetProcAddress(LoadLibraryA("Api-ms-win-core-version-l1-1-0.dll"), "VerQueryValueA");
+        VerQueryValueA(lpVI, "\\VarFileInfo\\Translation", (LPVOID*)&langInfo, &cbLang);
+
+        // Get Description
+        MSVCRT$sprintf(szVerDescription, "\\StringFileInfo\\%04x%04x\\%s", langInfo[0], langInfo[1], "FileDescription");
+        VerQueryValueA(lpVI, szVerDescription, &lpDescription, &cbBufSize);
+
+        // Get Company Name
+        MSVCRT$sprintf(szVerCompanyName, "\\StringFileInfo\\%04x%04x\\%s", langInfo[0], langInfo[1], "CompanyName");
+        VerQueryValueA(lpVI, szVerCompanyName, &lpCompanyName, &cbBufSize);
+
+        // Print line
+        internal_printf("%-50s%-25s%-25s\n", szFile, (LPTSTR)lpCompanyName, (LPTSTR)lpDescription);
+
+        //Cleanup
+        KERNEL32$GlobalFree((HGLOBAL)lpVI);
+    } else {
+        internal_printf("ERROR: Could not allocate memory\n");
+    }
+    return 0;
+}
+
+int PrintModules(DWORD processID)
+{
+    HMODULE hMods[256];     // if 1024 -> Unknown symbol '___chkstk_ms'
+    HANDLE hProcess;
+    DWORD cbNeeded;
+    unsigned int i;
+    char szModName[MAX_PATH];
+
+    // Print the process identifier. (debug)
+    //BeaconPrintf(CALLBACK_OUTPUT, "Process ID: %u\n", processID );
+
+    // Get a handle to the process.
+    hProcess = KERNEL32$OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, processID);
+    if (NULL == hProcess)
+        return 1;
+
+    if(PSAPI$EnumProcessModulesEx(hProcess, hMods, sizeof(hMods), &cbNeeded, LIST_MODULES_ALL)) {
```

Comment:
There is a chance that we miss modules with this setup.

To get all modules you would first call EnumProcessModulesEX passing NULL , 0 in place of hMods, sizeof(hMods) and that will give you the total size required.

Then you would use intAlloc(cbNeeded) to allocate a HMODULE array of that size.  This will also get around the __chkstk_ms issue.

After that you can make a second call to EnumProcessModulesEX and you will for sure get every module.

Reply:
Very good idea :) This behavior is not even documented lol. Thanks. It's fixed.

Reply:
You're welcome :)

I was also surprised the documentation wasn't more explicit about this.  From past experience almost every windows api that takes a DWORD * and outputs some size in bytes to it will have this feature to inform usermode applications how much memory they need to allocate to hold some arbitrary data.